iree/compiler/Dialect/Flow/IR/FlowOpUtils.cpp

```diff
@@ -14,131 +14,297 @@
+  static constexpr int kMaxInlinedConstantBytes = 256;
+
+  auto constantValueAttr = constantOp.getValue();
+  auto constantType = constantOp.getType();
+  if (constantValueAttr.isa<SplatElementsAttr>()) {
+    // Splats are always small and can often have special handling when we
+    // know they are a splat - which is why it's so important we inline them
+    // here so we know when they are used that's the case.
+    return true;
+  } else if (auto denseAttr = constantValueAttr.dyn_cast<DenseElementsAttr>()) {
+    // Smallish constants are worth moving inside.
+    auto shapedType = constantType.cast<ShapedType>();
+    uint64_t estimatedByteLength =
+        (shapedType.getNumElements() * shapedType.getElementTypeBitWidth()) / 8;
+    return denseAttr.isSplat() ||
+           estimatedByteLength <= kMaxInlinedConstantBytes;
+  } else if (constantType.isIntOrIndexOrFloat()) {
+    // Primitives can always go in.
+    return true;
+  }
+
+  return false;
+}
+
+// Returns true if the given value should be inlined into the closure region.
+// This is non-recursive and only holds for this value. Recursively cloning
+// trees is hard and it'd be better to model that differently such as by having
+// a wrapper region for immutable blobs that can be inlined that this then
+// returns true for.
+static bool shouldInlineIntoClosure(Value value) {
+  auto definingOp = value.getDefiningOp();
+  if (auto constantOp = dyn_cast<ConstantOp>(definingOp)) {
+    // Constants are perfect!
+    return isConstantSmall(constantOp);
+  } else if (auto variableLoadOp =
+                 dyn_cast<IREE::Flow::VariableLoadOp>(definingOp)) {
+    // If the variable is immutable then we can inline the reference to it.
+    auto variableOp =
+        SymbolTable::lookupNearestSymbolFrom<IREE::Flow::VariableOp>(
+            definingOp, variableLoadOp.variable());
+    return !variableOp.is_mutable();
+  }
+  return false;
+}
+
+// Inlines operands of the closure into the entry block as appropriate.
+// The closure operands and block arguments will remain untouched but all uses
+// will be replaced with the newly cloned values.
+//
+// Note that if multiple operands reference the same value it will get cloned
+// multiple times. That's fine, as anything we can inline here is something we
+// should also be able to CSE and that happens later on anyway.
+static void inlineClosureOperands(ClosureOpInterface &closureOp,
+                                  Block &entryBlock) {
+  auto builder = OpBuilder::atBlockBegin(&entryBlock);
+  for (auto opArg : llvm::enumerate(closureOp.getClosureOperands())) {
+    auto outerValue = opArg.value();
+    auto *sourceOp = outerValue.getDefiningOp();
+    if (!sourceOp) continue;  // can't clone block arguments into closures
+    if (closureOp.canClosureContainOp(sourceOp) &&
+        shouldInlineIntoClosure(outerValue)) {
+      // Clone the op (with regions).
+      auto *clonedOp = builder.clone(*sourceOp);
```

Comment:
how doyou ensure that the cloned op doesn't use values from outside the closure?

Reply:
Yeah this is tricky. In `DispatchLinalgOnTensors` have a whole amount logic to pull those in.

Reply:
Right now only constants are inlined and those don't have operands. I'll add a check here to indicate this is a restriction. We can extend this in the future but right now we can barely even pull in the constants so I think we're fine :)

Reply:
Oh cool - no check even needed here as I was already handling that as `shouldInlineIntoClosure` only picks constants and immutable variable loads. I even echo'ed both your sentiments in the comment :P 
```c
// This is non-recursive and only holds for this value. Recursively cloning
// trees is hard and it'd be better to model that differently such as by having
// a wrapper region for immutable blobs that can be inlined that this then
// returns true for.
```

The rematerialize dispatch constants pass had all the code to do recursive clones but in practice never did anything but splat constants, and I'm not sure any of that would even be needed in linalg on tensors world (or we should at least contain it there).